Cosmic new clustering

sophiemiddleton opened the pull request:
Merging in a few tidy-ups to the CosmicReco code before beginning Kalman Code. The eff. is worsened due to the new StrawResponse code but I think we must keep upto date so we will not have confusing histories. Also, I think Richie will do further improvements to the TimeClsutering so the New version here is a placeholder but does provide some benefits to efficiency.

FNALbuild commented:
Hi @sophiemiddleton,
You have proposed changes to files in these packages:
- Mu2eUtilities
- CosmicReco

which require these tests: build.




<a href="https://mu2ewiki.fnal.gov/wiki/Git#github_pull_request_procedures">cms-bot/mu2e commands are explained here</a>

ryuwd commented on CosmicReco/src/PDFFit.cc:
```diff
@@ -47,238 +54,234 @@ float strawradius = 2.5; //2.5 mm in mm
+	double a0 = x[0];
+	double a1 = x[1];
+	double b0 = x[2];
+	double b1 = x[3];
```
```suggestion
	double const& a0 = x[0];
	double const& a1 = x[1];
	double const& b0 = x[2];
	double const& b1 = x[3];
```
Left t0 alone because it gets modified later in the function

ryuwd commented on CosmicReco/src/PDFFit.cc:
```diff
@@ -47,238 +54,234 @@ float strawradius = 2.5; //2.5 mm in mm
+
+	double a0 = x[0];
+	double a1 = x[1];
+	double b0 = x[2];
```
```suggestion
	double const& a0 = x[0];
	double const& a1 = x[1];
	double const& b0 = x[2];
```
b1 too

kutschke commented:
@FNALbuild run build test

ryuwd commented:
Test - Compare line endings (sending from linux)
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/157/commits/f8f84f437c29dc84572e2cf74ad99cd318abdd15: build

FNALbuild commented:
:+1:
The build test passed at ref f8f84f437c29dc84572e2cf74ad99cd318abdd15.

For more details, please check [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//45/console).

ryuwd commented:
Just going to run a quick whitespace check, it does produce output, but you don't have to apply any of the output - just ignore for now.

@FNALbuild run code check

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/157/commits/f8f84f437c29dc84572e2cf74ad99cd318abdd15: code checks

FNALbuild commented:
:-1: Code formatting checks have failed at ref f8f84f437c29dc84572e2cf74ad99cd318abdd15.
#### fix-whitespace changed these files:
```
git diff --compact-summary

```

Please review and `git apply` [this patch](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//82/artifact/simple-format-pr157-f8f84f437c29dc84572e2cf74ad99cd318abdd15.patch) on your PR branch:
```
curl https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//82/artifact/simple-format-pr157-f8f84f437c29dc84572e2cf74ad99cd318abdd15.patch | git apply -v --index
git commit -am "Code formatting patch on f8f84f437c29dc84572e2cf74ad99cd318abdd15"
git push
```

To avoid this in future please do not use hard tabs to indent, and configure your editor to trim trailing whitespace.

sophiemiddleton commented:
I did the white space patch - looks a bit weird though

kutschke commented on CosmicReco/src/CosmicTrackMCInfo.cc:
```diff
@@ -1,8 +1,13 @@
+     TruePhiSIM = 0;
+     TrueThetaSIM = 0;
+     TrueMomentum = 0;
+
```
Two things:

1) These can be initialized in the class declaration ( in the .hh file)
        double TruePhiSIM = 0.;
    If something can be initialized that way we want everyone to do it.

2) If something cannot be initialized in the header, then use the intializer syntax ( ie the colon after the close ")" of the function arguments ).

Neither of these are big deal for this particular class but if you just do it everytime then you don't need to remember to do something special in the cases that do matter.

sophiemiddleton replied:
ah yes sure, sorry, will do that

kutschke commented:
Hi Sophie

  Have you committed the changes with the whitespace push?  If not, please hold off for now.

  What do you find weird about it?

Rob


> On Mar 14, 2020, at 8:39 AM, Sophie Middleton <notifications@github.com> wrote:
> 
> 
> I did the white space patch - looks a bit weird though
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

sophiemiddleton commented:
Hi, I reverted back the white space changes. It looked awful in the Github code-view pages, just indents all over the place, code not aligned etc.... I guess it did something strange in terms of counting number of tabs/spaces. I think it looks OK now in this format and in my editor, I guess the bot sees everything and its not specifically looking at the cosmetics of the code, rather the number of spaces and tabs. Ryun said the white space was optional so I reverted it, I can undo.

ryuwd commented:
Ah yes, apologies - I didn't expect the patch to be applied - I had a look through the patchfile and noticed it removed very large number of tab characters. 

At the moment, it doesn't replace hard tabs with spaces, it just removes them. That's probably why it looked bad.

kutschke commented:
HI Everyone,

  We definitely want to remove tab characters.  The question is how to replace them.

  Sophie: can you tell me the names of a few files to look at the SHAs for the before and after commits?

  Thanks,
                Rob

> On Mar 14, 2020, at 2:17 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Ah yes, apologies - I didn't expect the patch to be applied - I had a look through the patchfile and noticed it removed very large number of tab characters.
> 
> At the moment, it doesn't replace hard tabs with spaces, it just removes them. That's probably why it looked bad.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

brownd1978 commented:
I have tab set to 2 spaces in my editor.  Some people prefer 3.  There is
no 'right' answer.
I propose we analyze the existing code base and adopt the most
commonly-used convention (in terms of number of lines of code).

On Sat, Mar 14, 2020 at 4:10 PM Rob Kutschke <notifications@github.com>
wrote:

> HI Everyone,
>
> We definitely want to remove tab characters. The question is how to
> replace them.
>
> Sophie: can you tell me the names of a few files to look at the SHAs for
> the before and after commits?
>
> Thanks,
> Rob
>
> > On Mar 14, 2020, at 2:17 PM, ryuwd <notifications@github.com> wrote:
> >
> >
> > Ah yes, apologies - I didn't expect the patch to be applied - I had a
> look through the patchfile and noticed it removed very large number of tab
> characters.
> >
> > At the moment, it doesn't replace hard tabs with spaces, it just removes
> them. That's probably why it looked bad.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/157#issuecomment-599146134>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH5776552GYSWPINDODSLRHQFHFANCNFSM4LHHFPZQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

ryuwd commented:
Hi Rob,

You can see where the tab characters have been removed in this patchfile: https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//82/artifact/simple-format-pr157-f8f84f437c29dc84572e2cf74ad99cd318abdd15.patch

gaponenko commented:
> I have tab set to 2 spaces in my editor. Some people prefer 3. There is no 'right' answer. I propose we analyze the existing code base and adopt the most commonly-used convention (in terms of number of lines of code).

For C++ code one can replace all hard tabs with any number of spaces, then re-indent the code.  This works regardless of the original tab setting for the file.

ryuwd commented:
> > I have tab set to 2 spaces in my editor. Some people prefer 3. There is no 'right' answer. I propose we analyze the existing code base and adopt the most commonly-used convention (in terms of number of lines of code).
> 
> For C++ code one can replace all hard tabs with any number of spaces, then re-indent the code. This works regardless of the original tab setting for the file.

This is partially what I was hoping for with my [`fix-whitespace` script](https://github.com/ryuwd/codetools/blob/master/bin/fix-whitespace) - that any removed tab characters (assuming there weren't that many) would then be fixed manually or by using a re-indent script (e.g. gg=G in vim) .

I have changed the default behaviour to have hard tabs replaced with 4 spaces, rather than only removing the tabs.

FNALbuild commented:
:-1: Code formatting checks have failed at ref fd6a9d97804188e7f7256de81aa0dfdd43d28ba5.
#### fix-whitespace changed these files:
```
git diff --compact-summary
 CosmicReco/fcl/prolog.fcl                   |  254 ++---
 CosmicReco/inc/CosmicTrackFit.hh            |  112 +-
 CosmicReco/inc/CosmicTrackMCInfo.hh         |    2 +-
 CosmicReco/inc/DriftFitUtils.hh             |   24 +-
 CosmicReco/inc/MinuitDriftFitter.hh         |   32 +-
 CosmicReco/inc/PDFFit.hh                    |   66 +-
 CosmicReco/src/CosmicAnalyzer_module.cc     | 1506 +++++++++++++--------------
 CosmicReco/src/CosmicSeedFilter_module.cc   |   16 +-
 CosmicReco/src/CosmicTrackDetails_module.cc |  748 ++++++-------
 CosmicReco/src/CosmicTrackDiag_module.cc    |   56 +-
 CosmicReco/src/CosmicTrackFinder_module.cc  |  610 +++++------
 CosmicReco/src/CosmicTrackFit.cc            |  622 +++++------
 CosmicReco/src/CosmicTrackMCInfo.cc         |    6 +-
 CosmicReco/src/DriftFitUtils.cc             |  226 ++--
 CosmicReco/src/MinuitDriftFitter.cc         |  314 +++---
 CosmicReco/src/PDFFit.cc                    |  366 +++----
 CosmicReco/src/SimpleTimeCluster_module.cc  |    2 +-
 17 files changed, 2481 insertions(+), 2481 deletions(-)
```

Please review and `git apply` [this patch](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//83/artifact/simple-format-pr157-fd6a9d97804188e7f7256de81aa0dfdd43d28ba5.patch) on your PR branch:
```
curl https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//83/artifact/simple-format-pr157-fd6a9d97804188e7f7256de81aa0dfdd43d28ba5.patch | git apply -v --index
git commit -am "Code formatting patch on fd6a9d97804188e7f7256de81aa0dfdd43d28ba5"
git push
```

To avoid this in future please do not use hard tabs to indent, and configure your editor to trim trailing whitespace.

ryuwd commented:
The indentation is still inconsistent before and after the patch, but can be fixed by `clang-format` when desired:

```
# clean terminal environment
setup mu2e
setup clang v5_0_1
cd Offline
clang-format -i CosmicReco/inc/*.hh
clang-format -i CosmicReco/src/*.cc
```

kutschke commented:
Hi Ryun,

  What I want to see is before and after the patch so I can see what the effect of the patch is.   I could solve for what to apply the patch to and then do it - are the before and after available somewhere that I can see them.

Rob


> On Mar 14, 2020, at 6:25 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Hi Rob,
> 
> You can see where the tab characters have been removed in this patchfile: https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//82/artifact/simple-format-pr157-f8f84f437c29dc84572e2cf74ad99cd318abdd15.patch
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

ryuwd commented:
Hi Rob,

No problem, the formatting changes were applied on top of f8f84f437c29dc84572e2cf74ad99cd318abdd15, and committed at 3fb10ab388e3d332fcd83dc49d18051cd442f92f . The script removed leading hard tabs, and trailing whitespace:

For example, CosmicTrackFit.cc:

- Before: https://github.com/Mu2e/Offline/blob/f8f84f437c29dc84572e2cf74ad99cd318abdd15/CosmicReco/src/CosmicTrackFit.cc (i.e. `git checkout f8f84f437c29dc84572e2cf74ad99cd318abdd15`)
- After: https://github.com/Mu2e/Offline/blob/3fb10ab388e3d332fcd83dc49d18051cd442f92f/CosmicReco/src/CosmicTrackFit.cc (i.e. `git checkout 3fb10ab388e3d332fcd83dc49d18051cd442f92f`)

Best wishes, Ryun

kutschke commented:
Thanks Ryun,

  I agree with Sophie that this particular patch makes things worse.    

  I believe that this patch simply removes hard tabs without replacing them with anything - is that correct?

Rob


> On Mar 14, 2020, at 7:42 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Hi Rob,
> 
> No problem, the formatting changes were applied on top of f8f84f4, and committed at 3fb10ab . The script removed leading hard tabs, and trailing whitespace:
> 
> For example, CosmicTrackFit.cc:
> 
> 	• Before: https://github.com/Mu2e/Offline/blob/f8f84f437c29dc84572e2cf74ad99cd318abdd15/CosmicReco/src/CosmicTrackFit.cc (i.e. git checkout f8f84f437c29dc84572e2cf74ad99cd318abdd15)
> 	• After: https://github.com/Mu2e/Offline/blob/3fb10ab388e3d332fcd83dc49d18051cd442f92f/CosmicReco/src/CosmicTrackFit.cc (i.e. git checkout 3fb10ab388e3d332fcd83dc49d18051cd442f92f)
> Best wishes, Ryun
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

ryuwd commented:
Hi Rob,

Yes, you're right - it removed hard tabs completely, and didn't replace them with spaces, leading to the messed-up indentation you saw in the comparison.

I have since changed the script to replace leading hard tabs with 4 spaces. 

The results have improved. Here is the new patch file that FNALbuild generated:
https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//83/artifact/simple-format-pr157-fd6a9d97804188e7f7256de81aa0dfdd43d28ba5.patch

Here is a snippet of the diff from PDFFit.hh (If the diff is difficult to read, try reading it on the GitHub website.)
```diff
diff --git a/CosmicReco/inc/PDFFit.hh b/CosmicReco/inc/PDFFit.hh
index d0588838d..ea7bfe34b 100644
--- a/CosmicReco/inc/PDFFit.hh
+++ b/CosmicReco/inc/PDFFit.hh
@@ -22,48 +22,48 @@
 using namespace mu2e;
 
     class GaussianPDFFit : public ROOT::Minuit2::FCNBase {
-	public:
-		ComboHitCollection chits;
-		StrawResponse const& srep ;
-		CosmicTrack track;
-		std::vector<double> docas;
+    public:
+        ComboHitCollection chits;
+        StrawResponse const& srep ;
+        CosmicTrack track;
+        std::vector<double> docas;
 
-		std::vector<double> constraint_means;
-		std::vector<double> constraints;
-		double sigma_t;
-		int k;
+        std::vector<double> constraint_means;
+        std::vector<double> constraints;
+        double sigma_t;
+        int k;
 
-		const Tracker* tracker;
-		std::vector<double> DOCAs;
-		std::vector<double> TimeResiduals;
+        const Tracker* tracker;
+        std::vector<double> DOCAs;
+        std::vector<double> TimeResiduals;
 
-		int nparams =5; 
+        int nparams =5;
 
-		GaussianPDFFit(ComboHitCollection _chits, StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double _sigma_t, int _k,  const Tracker* _tracker) :  chits(_chits),  srep(_srep), track(_track), constraint_means(_constraint_means), constraints(_constraints) , sigma_t(_sigma_t), k(_k), tracker(_tracker) {};
+        GaussianPDFFit(ComboHitCollection _chits, StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double _sigma_t, int _k,  const Tracker* _tracker) :  chits(_chits),  srep(_srep), track(_track), constraint_means(_constraint_means), constraints(_constraints) , sigma_t(_sigma_t), k(_k), tracker(_tracker) {};
 
-		double Up() const { return 0.5; };
-		double operator() (const std::vector<double> &x) const;
-		double TimeResidual(double doca, StrawResponse const& srep , double t0, ComboHit const& hit,  const Tracker* tracker) const ;
-		double calculate_DOCA(ComboHit const& chit, double a0, double a1, double b0, double b1, const Tracker* tracker) const;
-		double calculate_ambig(ComboHit const& chit, double a0, double a1, double b0, double b1,  const Tracker* tracker) const;
+        double Up() const { return 0.5; };
+        double operator() (const std::vector<double> &x) const;
+        double TimeResidual(double doca, StrawResponse const& srep , double t0, ComboHit const& hit,  const Tracker* tracker) const ;
+        double calculate_DOCA(ComboHit const& chit, double a0, double a1, double b0, double b1, const Tracker* tracker) const;
+        double calculate_ambig(ComboHit const& chit, double a0, double a1, double b0, double b1,  const Tracker* tracker) const;
 
     };
 
     class FullDriftFit : public GaussianPDFFit {
-    	public:
-		FullDriftFit(ComboHitCollection _chits,  StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double sigma_t, int _k, const Tracker* _tracker);
-		int Factorial(int k);
-		void CalculateFullPDF();
-		double InterpolatePDF(double time_residual, double sigma, double tau) const;
-		void DeleteArrays() const;
-		double Min_t;
-		double  Min_tau, Min_s;
-		double Max_t, Max_tau, Max_s;
-		double delta_T, delta_Tau, delta_S;
-		double *pdf_sigmas, *pdf_taus, *pdf_times;
-		double *pdf;
-		int k;
-		double operator() (const std::vector<double> &x) const;
+        public:
+        FullDriftFit(ComboHitCollection _chits,  StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double sigma_t, int _k, const Tracker* _tracker);
+        int Factorial(int k);
+        void CalculateFullPDF();
+        double InterpolatePDF(double time_residual, double sigma, double tau) const;
+        void DeleteArrays() const;
+        double Min_t;
+        double  Min_tau, Min_s;
+        double Max_t, Max_tau, Max_s;
+        double delta_T, delta_Tau, delta_S;
+        double *pdf_sigmas, *pdf_taus, *pdf_times;
+        double *pdf;
+        int k;
+        double operator() (const std::vector<double> &x) const;
 };
 
 #endif
```

kutschke commented:
Hi Ryun,

  The only thing that makes sense is to remove hard tabs and then to run the standard formatter.  So the group needs to decide on a standard format.  

    Just replacing tabs with N spaces can never work.   If someone has "space tab", replacing it with a fixed number of spaces leaves the result off by one.    The original physical behaviour of a typewritter was that a tab advanced from the current position to the next predefined tab stop but I don't think that gives the behaviour we want either.

   Thanks

        Rob


> On Mar 14, 2020, at 10:57 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Hi Rob,
> 
> Yes, you're right - it removed hard tabs completely, and didn't replace them with spaces, leading to the messed-up indentation you saw in the comparison.
> 
> I have since changed the script to replace leading hard tabs with 4 spaces.
> 
> The results have improved. Here is the new patch file that FNALbuild generated:
> https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//83/artifact/simple-format-pr157-fd6a9d97804188e7f7256de81aa0dfdd43d28ba5.patch
> 
> Here is a snippet of the diff from PDFFit.hh (If the diff is difficult to read, try reading it on the GitHub website.)
> 
> diff --git a/CosmicReco/inc/PDFFit.hh b/CosmicReco/inc/PDFFit.hh
> 
> index d0588838d..ea7bfe34b 100644
> 
> --- a/CosmicReco/inc/PDFFit.hh
> +++ b/CosmicReco/inc/PDFFit.hh
> @@ -22,48 +22,48 @@
> 
>  using namespace mu2e;
>  
>      class GaussianPDFFit : public ROOT::Minuit2::FCNBase {
> 
> -	public:
> -		ComboHitCollection chits;
> -		StrawResponse const& srep ;
> -		CosmicTrack track;
> -		std::vector<double> docas;
> +    public:
> +        ComboHitCollection chits;
> +        StrawResponse const& srep ;
> +        CosmicTrack track;
> +        std::vector<double> docas;
> 
>  
> 
> -		std::vector<double> constraint_means;
> -		std::vector<double> constraints;
> -		double sigma_t;
> -		int k;
> +        std::vector<double> constraint_means;
> +        std::vector<double> constraints;
> +        double sigma_t;
> +        int k;
> 
>  
> 
> -		const Tracker* tracker;
> -		std::vector<double> DOCAs;
> -		std::vector<double> TimeResiduals;
> +        const Tracker* tracker;
> +        std::vector<double> DOCAs;
> +        std::vector<double> TimeResiduals;
> 
>  
> 
> -		int nparams =5; 
> +        int nparams =5;
> 
>  
> 
> -		GaussianPDFFit(ComboHitCollection _chits, StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double _sigma_t, int _k,  const Tracker* _tracker) :  chits(_chits),  srep(_srep), track(_track), constraint_means(_constraint_means), constraints(_constraints) , sigma_t(_sigma_t), k(_k), tracker(_tracker) {};
> +        GaussianPDFFit(ComboHitCollection _chits, StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double _sigma_t, int _k,  const Tracker* _tracker) :  chits(_chits),  srep(_srep), track(_track), constraint_means(_constraint_means), constraints(_constraints) , sigma_t(_sigma_t), k(_k), tracker(_tracker) {};
> 
>  
> 
> -		double Up() const { return 0.5; };
> -		double operator() (const std::vector<double> &x) const;
> -		double TimeResidual(double doca, StrawResponse const& srep , double t0, ComboHit const& hit,  const Tracker* tracker) const ;
> -		double calculate_DOCA(ComboHit const& chit, double a0, double a1, double b0, double b1, const Tracker* tracker) const;
> -		double calculate_ambig(ComboHit const& chit, double a0, double a1, double b0, double b1,  const Tracker* tracker) const;
> +        double Up() const { return 0.5; };
> +        double operator() (const std::vector<double> &x) const;
> +        double TimeResidual(double doca, StrawResponse const& srep , double t0, ComboHit const& hit,  const Tracker* tracker) const ;
> +        double calculate_DOCA(ComboHit const& chit, double a0, double a1, double b0, double b1, const Tracker* tracker) const;
> +        double calculate_ambig(ComboHit const& chit, double a0, double a1, double b0, double b1,  const Tracker* tracker) const;
> 
>  
>      };
>  
>      class FullDriftFit : public GaussianPDFFit {
> 
> -    	public:
> -		FullDriftFit(ComboHitCollection _chits,  StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double sigma_t, int _k, const Tracker* _tracker);
> -		int Factorial(int k);
> -		void CalculateFullPDF();
> -		double InterpolatePDF(double time_residual, double sigma, double tau) const;
> -		void DeleteArrays() const;
> -		double Min_t;
> -		double  Min_tau, Min_s;
> -		double Max_t, Max_tau, Max_s;
> -		double delta_T, delta_Tau, delta_S;
> -		double *pdf_sigmas, *pdf_taus, *pdf_times;
> -		double *pdf;
> -		int k;
> -		double operator() (const std::vector<double> &x) const;
> +        public:
> +        FullDriftFit(ComboHitCollection _chits,  StrawResponse const& _srep, CosmicTrack _track, std::vector<double> &_constraint_means, std::vector<double> &_constraints, double sigma_t, int _k, const Tracker* _tracker);
> +        int Factorial(int k);
> +        void CalculateFullPDF();
> +        double InterpolatePDF(double time_residual, double sigma, double tau) const;
> +        void DeleteArrays() const;
> +        double Min_t;
> +        double  Min_tau, Min_s;
> +        double Max_t, Max_tau, Max_s;
> +        double delta_T, delta_Tau, delta_S;
> +        double *pdf_sigmas, *pdf_taus, *pdf_times;
> +        double *pdf;
> +        int k;
> +        double operator() (const std::vector<double> &x) const;
> 
>  };
>  
>  #endif
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

ryuwd commented:
Hi Rob,

In that case perhaps the best thing would be to run `clang-format`. This will re-indent each file, making sure the indent size is 4 spaces throughout, and no hard tabs should remain.

(It's probably best to do this after the non-formatting related changes have been approved)

```bash
# in a new terminal window
setup mu2e
setup clang v5_0_1

# Uses the config file merged in at #144.
clang-format -i CosmicReco/src/*.cc
clang-format -i CosmicReco/inc/*.hh
git commit -am "clang-format on CosmicReco"
git push
```

sophiemiddleton commented:
Let me know what the descision is with regards to this PR

kutschke commented:
Hi Ryun,

 Following up on Dave's suggestion.  Can we use clang tidy to let us know what the most common indenting style is?

Rob


> On Mar 15, 2020, at 1:24 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Hi Rob,
> 
> In that case perhaps the best thing would be to run clang-format. This will re-indent each file, making sure the indent size is 4 spaces throughout, and no hard tabs should remain.
> 
> (It's probably best to do this after the non-formatting related changes have been approved)
> 
> # in a new terminal window
> 
> setup mu2e
> setup clang v5_0_1
> 
> 
> # Uses the config file merged in at #144.
> 
> clang-format -i CosmicReco/src/
> *
> .cc
> clang-format -i CosmicReco/inc/
> *
> .hh
> git commit -am 
> "clang-format on CosmicReco"
> 
> git push
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

ryuwd commented:
Hi Rob, 

I had a quick sweep through the clang-format and clang-tidy documentation and unfortunately there's no option to determine the most common indentation of a file. 

I wrote a short script to determine the most common indent size and ran it over all .cc and .hh files in Offline: https://gist.github.com/ryuwd/40f9138fdfe59191c9cd4cdb55361006

The results:

```
{
0: 112550, # n leading spaces in line not divisible by indent block with zero remainder
 -1: 16702, #  not an indent size of 2, 3, 4, or 8

2: 91234, # n lines
3: 7172, # n lines
4: 9604, # n lines
8: 903 # n lines
}
```

Best wishes, Ryun

kutschke commented:
Hi Ryun,

  Are your numbers ident relative to column 0?  Indent relative to previous line? Indent relative to previous non-blank line?
Indent relative to previous open brace?   Something else?  Do you suppress lines that are comments?  

  Rob





> On Mar 15, 2020, at 3:07 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Hi Rob,
> 
> I had a quick sweep through the clang-format and clang-tidy documentation and unfortunately there's no option to determine the most common indentation of a file.
> 
> I wrote a short script to determine the most common indent size and ran it over all .cc and .hh files in Offline: https://gist.github.com/ryuwd/40f9138fdfe59191c9cd4cdb55361006
> 
> The results:
> 
> {
> 0: 112550, # n leading spaces in line not divisible by indent block with zero remainder
>  -1: 16702, #  not an indent size of 2, 3, 4, or 8
> 
> 2: 91234, # n lines
> 3: 7172, # n lines
> 4: 9604, # n lines
> 8: 903 # n lines
> }
> 
> Best wishes, Ryun
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

ryuwd commented:
Hi Rob,

*Hi Ryun, *
>
> * Are your numbers ident relative to column 0? Indent relative to previous
> line? Indent relative to previous non-blank line? *


I deduce which 'indent block' we're in by looking for opening and closing
braces. When it finds '{' in a line, it increases the indent_block variable
by 1. When it sees '}' it decreases by one.
It then counts how many spaces lead up to the first non-whitespace
character in a line, then divides by indent_block to find the number of
spaces per indentation.

It doesn't count the line if:

   - n_spaces / indent_block has a remainder
   - there were tabs in the leading whitespace
   - the indent size wasn't 2, 3, 4, or 8



> * Indent relative to previous open brace? Something else? Do you suppress
> lines that are comments? *

At the moment it does count lines with comments in them.

Best wishes, Ryun

On Sun, 15 Mar 2020 at 21:06, Rob Kutschke <notifications@github.com> wrote:

> Hi Ryun,
>
> Are your numbers ident relative to column 0? Indent relative to previous
> line? Indent relative to previous non-blank line?
> Indent relative to previous open brace? Something else? Do you suppress
> lines that are comments?
>
> Rob
>
>
>
>
>
> > On Mar 15, 2020, at 3:07 PM, ryuwd <notifications@github.com> wrote:
> >
> >
> > Hi Rob,
> >
> > I had a quick sweep through the clang-format and clang-tidy
> documentation and unfortunately there's no option to determine the most
> common indentation of a file.
> >
> > I wrote a short script to determine the most common indent size and ran
> it over all .cc and .hh files in Offline:
> https://gist.github.com/ryuwd/40f9138fdfe59191c9cd4cdb55361006
> >
> > The results:
> >
> > {
> > 0: 112550, # n leading spaces in line not divisible by indent block with
> zero remainder
> > -1: 16702, # not an indent size of 2, 3, 4, or 8
> >
> > 2: 91234, # n lines
> > 3: 7172, # n lines
> > 4: 9604, # n lines
> > 8: 903 # n lines
> > }
> >
> > Best wishes, Ryun
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/157#issuecomment-599264651>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANOMGYXI5ZSAORFH2ZUE7NDRHU7MZANCNFSM4LHHFPZQ>
> .
>

kutschke commented:
Hi Ryun,

  In a c'tor intiailizer list, clang-format is collapsing multiple lines with a single initialization onto one line.  For example:

-    _hStrawEDep(0),
-    _hNstraw(0),
-    _hCrystalEDep(0),
-    _hNcrystal(0),
-    _ntup(0)
+    _hStrawEDep(0), _hNstraw(0), _hCrystalEDep(0), _hNcrystal(0), _ntup(0)

Is there  away to tell clang-format to respect what the author did?  Or to force one per line?

Aside: in this case it's moot.  This code was written ages ago.  The next time I touch it I will initializer these in the declaration.

  Also, how to do change:

  void function()
  {
  }

  To 

  void function(){
  }

 Thanks,

        Rob




> On Mar 15, 2020, at 4:25 PM, ryuwd <notifications@github.com> wrote:
> 
> 
> Hi Rob,
> 
> *Hi Ryun, *
> >
> > * Are your numbers ident relative to column 0? Indent relative to previous
> > line? Indent relative to previous non-blank line? *
> 
> 
> I deduce which 'indent block' we're in by looking for opening and closing
> braces. When it finds '{' in a line, it increases the indent_block variable
> by 1. When it sees '}' it decreases by one.
> It then counts how many spaces lead up to the first non-whitespace
> character in a line, then divides by indent_block to find the number of
> spaces per indentation.
> 
> It doesn't count the line if:
> 
> - n_spaces / indent_block has a remainder
> - there were tabs in the leading whitespace
> - the indent size wasn't 2, 3, 4, or 8
> 
> 
> 
> > * Indent relative to previous open brace? Something else? Do you suppress
> > lines that are comments? *
> 
> At the moment it does count lines with comments in them.
> 
> Best wishes, Ryun
> 
> On Sun, 15 Mar 2020 at 21:06, Rob Kutschke <notifications@github.com> wrote:
> 
> > Hi Ryun,
> >
> > Are your numbers ident relative to column 0? Indent relative to previous
> > line? Indent relative to previous non-blank line?
> > Indent relative to previous open brace? Something else? Do you suppress
> > lines that are comments?
> >
> > Rob
> >
> >
> >
> >
> >
> > > On Mar 15, 2020, at 3:07 PM, ryuwd <notifications@github.com> wrote:
> > >
> > >
> > > Hi Rob,
> > >
> > > I had a quick sweep through the clang-format and clang-tidy
> > documentation and unfortunately there's no option to determine the most
> > common indentation of a file.
> > >
> > > I wrote a short script to determine the most common indent size and ran
> > it over all .cc and .hh files in Offline:
> > https://gist.github.com/ryuwd/40f9138fdfe59191c9cd4cdb55361006
> > >
> > > The results:
> > >
> > > {
> > > 0: 112550, # n leading spaces in line not divisible by indent block with
> > zero remainder
> > > -1: 16702, # not an indent size of 2, 3, 4, or 8
> > >
> > > 2: 91234, # n lines
> > > 3: 7172, # n lines
> > > 4: 9604, # n lines
> > > 8: 903 # n lines
> > > }
> > >
> > > Best wishes, Ryun
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub, or unsubscribe.
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Mu2e/Offline/pull/157#issuecomment-599264651>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/ANOMGYXI5ZSAORFH2ZUE7NDRHU7MZANCNFSM4LHHFPZQ>
> > .
> >
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

kutschke commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/157/commits/fd6a9d97804188e7f7256de81aa0dfdd43d28ba5: build

FNALbuild commented:
:+1:
The build test passed at ref fd6a9d97804188e7f7256de81aa0dfdd43d28ba5.

For more details, please check [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//48/console).

kutschke commented:
> Let me know what the descision is with regards to this PR

Sophie:  Ryun and I hi-jacked your pull request - my apologies.

We can finish the formatted discussion outside of this pull request. I re-ran the build.  I am approving it and merging it now.

sophiemiddleton commented:
Its fine, these are things which need planning. I will try to improve my tab/space practice for future work. Thanks, Sophie
